R/marshal.R

```diff
@@ -0,0 +1,131 @@
+#' * the function `marshaled_model(model)`, which returns `TRUE` if the model inherits from class `"marshaled"`
+#'   and `FALSE` otherwise.
+#'
+#' In order to implement marshaling for a Learner, you only need to overload the `marshal_model` and `unmarshal_model`
+#' methods and tag the learner with the `"marshal"` property accordingly.
+#'
+#' To make marshaling accessible in an R6-manner, you should also add the public methods `$marshal()`, `$unmarshal()`
+#' and the active binding `$marshaled`.
+#' To make this as convenient as possible, the functions `learner_marshal(learner)`, `learner_unmarshal(learner)`
+#' and `learner_marshaled(learner)` are provided and can be called from within the public methods.
+#' All three functions throw an error if the learner is not trained and otherwise call
+#' `marshal_model()`, `unmarshal_model()` or `marshaled_model()` on the learner's model.
+#'
+#' You can verify whether you have correctly implemented marshaling by using the internal test helper
+#' `expect_marshalable_learner()`. This is also run by `expect_learner()` if a task is provided.
+#'
+#' For a concrete example on how to implement marshaling, see [`LearnerClassifLily`].
+#'
+#' @param learner [`Learner`]\cr
+#'   The learner.
+#' @keywords internal
+#' @export
+learner_unmarshal = function(learner) {
+  # no need to check for 'marshal' property as this method should only be available for such learners
+  if (is.null(learner$model)) {
+    stopf("Cannot unmarshal, Learner '%s' has not been trained yet", learner$id)
+  }
+  # this will do nothing if the model was not marshaled
+  learner$model = unmarshal_model(learner$model)
+  invisible(learner)
+}
+
+#' @rdname marshaling
+#' @export
+learner_marshal = function(learner) {
+  # no need to check for 'marshal' property as this method should only be available for such learners
+  if (is.null(learner$model)) {
+    stopf("Cannot marshal, Learner '%s' has not been trained yet", learner$id)
+  }
+  # this will do nothing if the model was already marshaled
+  learner$model = marshal_model(learner$model)
+  invisible(learner)
+}
+
+#' @rdname marshaling
+#' @export
+learner_marshaled = function(learner) {
+  # no need to check for 'marshal' property as this method should only be available for such learners
+  if (is.null(learner$model)) {
+    stopf("Cannot check marshaled status, Learner '%s' has not been trained yet", learner$id)
+  }
+  marshaled_model(learner$model)
+}
+
+#' @rdname marshaling
+#' @export
+marshal_model = function(model, ...) {
+  UseMethod("marshal_model")
+}
+
+#' @rdname marshaling
+#' @export
+unmarshal_model = function(model, ...) {
+  if (marshaled_model(model) && is.character(model$packages)) {
+    require_namespaces(model$packages)
+  }
+  UseMethod("unmarshal_model")
+}
+
+#' @rdname marshaling
+#' @export
+marshaled_model = function(model) {
+  test_class(model, "marshaled")
+}
+
+#' @export
+marshal_model.default = function(model, ...) {
```

Comment:
remove this method as it is never called.

Reply:
do we really need the 'marshal' property.

Reply:
is marshaling defined as `unmarshal(marshal(x))` returns `x` or does it also include that `marshaled(marshal(x))` returns `TRUE`.